Move room/call API to central location

fancycode opened the pull request:
This is a work in progress to get some early feedback. Will need more changes, especially with #349 and while testing and thinking about how the final API should look like.

The goal is to be able to replace the whole "what-rooms-exist", "who-is-which-room" and "who-is-in-a-call" functionality with a different implementation at some point.

nickvergessen commented on js/app.js:
```diff
@@ -363,16 +349,17 @@
 			}
 		},
+		addParticipantToCall: function(token, participant) {
+			return this.signaling.addParticipantToCall(token, participant);
```
this should be were `createPublicVideoCall` is as well?
To make JS and php similar ( #349 )

It has nothing to do with signaling from my pov?

fancycode replied:
It is related to signaling as the signaling backend keeps track of who is in which call.

nickvergessen replied:
Not really, the room management decides whether a person is in a call or not?

fancycode replied:
Well, that would require that all clients poll the (slow) PHP regularly about who is in the call and we want to avoid that by moving this to the standalone signaling server. The PHP backend is only used to check which rooms / calls exist.

nickvergessen replied:
Okay and when a moderator kicks a user from a room we send a request to signaling and the room management then?

Basically need to do the same when adding users, because the data must be persistent?

fancycode replied:
Yeah, the database has a (persistent) list of users invited to a room. This is queried by the signaling when a user wants to join. The list of currently connected users in a room is kept by the signaling server (non-persistent).

Once a user is no longer invited (i.e. gets kicked), a notification must be sent from the db to the signaling to close any active connections of that user (to the room she is no longer invited to).

nickvergessen replied:
just to be sure, this route (addParticipant) is to invite a user to a room, not when a user joins/loads a room. That's why I think it should be in the other one, and the php backend then drops a similar message to the signaling server like on "kick"

fancycode replied:
Ah ok, if it's the invitation then it shouldn't go through the signaling. I'll change it back to go to the PHP backend.

nickvergessen commented on js/app.js:
```diff
@@ -705,5 +686,5 @@
 
 $(window).unload(function () {
-	OCA.SpreedMe.Rooms.leaveAllRooms();
+	OCA.SpreedMe.Calls.leaveAllRooms();
```
I think this should be leaveAllCalls, not leaveAllRooms.

nickvergessen commented on js/views/roomlistview.js:
```diff
@@ -292,22 +292,23 @@
 				OC.Util.History.pushState({}, OC.generateUrl('/apps/spreed'));
+			} else {
+				// TODO(fancycode): When does this happen? The user is in a call
+				// where the call is not active?
```
The user could be in a different call, then the call that is being left/deleted is not the active one. But I think this is theory only, because there is no UI to leave a call which is not active atm.

fancycode replied:
Ok, so I'll leave the else block with the comment for now.

nickvergessen commented on js/views/roomlistview.js:
```diff
@@ -292,22 +292,23 @@
-				OCA.SpreedMe.Rooms.leaveCurrentRoom();
-				OCA.SpreedMe.Rooms.showRoomDeletedMessage(true);
+				OCA.SpreedMe.Calls.leaveCurrentCall(true);
 				OC.Util.History.pushState({}, OC.generateUrl('/apps/spreed'));
```
this is done in `leaveCurrentCall` already,
but the ajax call is missing, because you moved it into the else...

fancycode replied:
Thanks for spotting, removed here as it's done in `leaveCurrentCall`.

nickvergessen commented on js/signaling.js:
```diff
@@ -41,13 +43,79 @@
+	SignalingBase.prototype.createOneToOneVideoCall = function(recipientUserId) {
+		var defer = $.Deferred();
+		$.ajax({
+			url: OC.linkToOCS('apps/spreed/api/v1', 2) + 'oneToOne',
```
we didn't have this merged @fancycode 
shall I rebase it for you?

fancycode replied:
I already have some more changes locally, so I will rebase myself, thanks.

fancycode commented:
Rebased and updated so the whole room pinging code is now part of the `InternalSignaling` class. Other changes in preparation to replace the signaling server backend.

@nickvergessen @Ivansss ptal

nickvergessen commented on js/views/roomlistview.js:
```diff
@@ -524,7 +524,7 @@
 				var token = _this.model.get('token');
 				var participant = e.val;
-				OCA.SpreedMe.app.addParticipantToCall(token, participant);
+				OCA.SpreedMe.app.addParticipantToRoom(token, participant);
```
😉

nickvergessen commented on js/signaling.js:
```diff
@@ -99,57 +131,119 @@
+		this._stopPingRoom();
+		if (this.currentCallToken) {
+			$.ajax({
+				url: OC.linkToOCS('apps/spreed/api/v1/room', 2) + this.currentCallToken + '/participants/self',
```
same here, this can also be done while not in a call. Just click the dot-menu on the room list and choose "leave room/call"

nickvergessen replied:
hence it is in the room api and not call

fancycode replied:
Hmm, I see. Is there a way to leave a call and stay in the room?

nickvergessen replied:
Yeah it's `leaveAllRooms` which should be named `leaveAllCalls`

nickvergessen commented on js/views/roomlistview.js:
```diff
@@ -290,22 +290,23 @@
+				// TODO(fancycode): When does this happen? The user is in a call
+				// where the call is not active?
+				$.ajax({
+					url: OC.linkToOCS('apps/spreed/api/v1/room', 2) + this.model.get('token') + '/participants/self',
```
this removes you from the room list, not from the call, so it should also work when "not active"
and the ajax call is missing from the "active" part of the if?

fancycode replied:
The ajax call is in `leaveCall` in the signaling code (you just commented on that), which is triggered by SimpleWebRTC when a user leaves a call. Again probably some room vs. call confusion on my side ;-)

nickvergessen commented on js/signaling.js:
```diff
@@ -99,57 +131,119 @@
+
+	InternalSignaling.prototype.leaveAllRooms = function() {
+		$.ajax({
+			url: OC.linkToOCS('apps/spreed/api/v1/call', 2) + this.currentCallToken,
```
since it now has the token and is not room related but call, it should be `leaveCall` not `leaveAllRooms`

fancycode commented:
I did some more changes for room vs. call, should look better now. Thanks for your feedback so far.

fancycode commented:
This should be good to go now.

@nickvergessen @Ivansss ptal

nickvergessen commented:
Conflicting files
js/views/roomlistview.js

fancycode commented:
Rebased

Ivansss commented:
There is still one eslint error:
```
/drone/src/github.com/nextcloud/spreed/pull/351/js/signaling.js
223:59 error 'rooms' is defined but never used no-unused-vars
```

fancycode commented:
Thanks, fixed now.
